File: NodeDesignPatterns/observer.js
"use strict";
let doc = `

                +--------------------+
                |      Subject       |
                +--------------------+
                | - observers: list  |
                +--------------------+
                | + attach(obs)      |
                | + detach(obs)      |
                | + notify()         |
                +--------------------+
                         |
                         |
                         v
        -------------------------------------
        |                 |                 |
+----------------+ +----------------+ +----------------+
|   Observer     | |   Observer     | |   Observer     |
+----------------+ +----------------+ +----------------+
| + update()     | | + update()     | | + update()     |
+----------------+ +----------------+ +----------------+
        ^                 ^                 ^
        |                 |                 |
+----------------+ +----------------+ +----------------+
| ConcreteObsA   | | ConcreteObsB   | | ConcreteObsC   |
+----------------+ +----------------+ +----------------+
| + update()     | | + update()     | | + update()     |
+----------------+ +----------------+ +----------------+


`;
console.log(doc);
class Mobile {
    update(stock, price) {
        console.log(`MobileApp: ${stock} price updated to ${price}`);
    }
}
class Webdashboard {
    update(stock, price) {
        console.log(`Webdashboard: ${stock} price updated to ${price}`);
    }
}
class Stock {
    stock_name;
    price;
    observer;
    constructor(stock) {
        this.stock_name = stock;
        this.price = 0;
        this.observer = [];
    }
    attach(observer) {
        this.observer.push(observer);
    }
    deattach(observer) {
        this.observer = this.observer.filter(obs => obs !== observer);
    }
    notify() {
        for (let i of this.observer) {
            i.update(this.stock_name, this.price);
        }
    }
    set_price(price) {
        this.price = price;
        this.notify();
    }
}
let s = new Stock("TCS");
s.attach(new Mobile());
s.attach(new Webdashboard());
s.set_price(100);

File: PythonDesignPatterns/observer.py
"""

                +--------------------+
                |      Subject       |
                +--------------------+
                | - observers: list  |
                +--------------------+
                | + attach(obs)      |
                | + detach(obs)      |
                | + notify()         |
                +--------------------+
                         |
                         |
                         v
        -------------------------------------
        |                 |                 |
+----------------+ +----------------+ +----------------+
|   Observer     | |   Observer     | |   Observer     |
+----------------+ +----------------+ +----------------+
| + update()     | | + update()     | | + update()     |
+----------------+ +----------------+ +----------------+
        ^                 ^                 ^
        |                 |                 |
+----------------+ +----------------+ +----------------+
| ConcreteObsA   | | ConcreteObsB   | | ConcreteObsC   |
+----------------+ +----------------+ +----------------+
| + update()     | | + update()     | | + update()     |
+----------------+ +----------------+ +----------------+



"""


print(__doc__)

from abc import ABC, abstractmethod

class Observer(ABC):

    @abstractmethod
    def update(self,stock_name,price):
        pass



class Stock:

    def __init__(self,stock):
        self.stock=stock
        self.price=0
        self.observer=[]
    
    def attach(self,observer):
        self.observer.append(observer)
    def deattach(self,observer):
        self.observer.remove(observer)

    def notify(self):
        for i in self.observer:
            i.update(self.stock,self.price)
    
    def set_price(self,price):
        self.price=price
        self.notify()



class Mobile(Observer):

    def update(self, stock_name, price):
        print(f"MobileApp: {stock_name} price updated to {price}")

class Webdashboard(Observer):

    def update(self, stock_name, price):
        print(f"webdashboard: {stock_name} price updated to {price}")


s=Stock("TCS")
s.attach(Mobile())
s.attach(Webdashboard())
s.set_price(100)